src/DotNetWorker.Core/Hosting/ServiceCollectionExtensions.cs

```diff
@@ -41,52 +41,47 @@ public static IFunctionsWorkerApplicationBuilder AddFunctionsWorkerCore(this ISe
 
             // Invocation Features
             services.TryAddSingleton<IInvocationFeaturesFactory, DefaultInvocationFeaturesFactory>();
-            services.AddSingleton<IInvocationFeatureProvider, DefaultBindingFeatureProvider>();
+            services.TryAddSingleton<IInvocationFeatureProvider, DefaultBindingFeatureProvider>();
 
             // Input conversion feature
-            services.AddSingleton<IConverterContextFactory, DefaultConverterContextFactory>();
-            services.AddSingleton<IInputConversionFeatureProvider, DefaultInputConversionFeatureProvider>();
-            services.AddSingleton<IInputConverterProvider, DefaultInputConverterProvider>();
+            services.TryAddSingleton<IConverterContextFactory, DefaultConverterContextFactory>();
+            services.TryAddSingleton<IInputConversionFeatureProvider, DefaultInputConversionFeatureProvider>();
+            services.TryAddSingleton<IInputConverterProvider, DefaultInputConverterProvider>();
 
             // Input binding cache
-            services.AddScoped<IBindingCache<ConversionResult>, DefaultBindingCache<ConversionResult>>();
+            services.TryAddScoped<IBindingCache<ConversionResult>, DefaultBindingCache<ConversionResult>>();
 
             // Output Bindings
-            services.AddSingleton<IOutputBindingsInfoProvider, DefaultOutputBindingsInfoProvider>();
+            services.TryAddSingleton<IOutputBindingsInfoProvider, DefaultOutputBindingsInfoProvider>();
 
             // Worker initialization service
-            services.AddSingleton<IHostedService, WorkerHostedService>();
+            services.TryAddSingleton<IHostedService, WorkerHostedService>();
 
             // Default serializer settings
-            services.AddOptions<WorkerOptions>()
-                .PostConfigure<IOptions<JsonSerializerOptions>>((workerOptions, serializerOptions) =>
-                {
-                    if (workerOptions.Serializer is null)
-                    {
-                        workerOptions.Serializer = new JsonObjectSerializer(serializerOptions.Value);
-                    }
-                });
+            services.AddOptions();
+            services.TryAddEnumerable(ServiceDescriptor.Transient<IPostConfigureOptions<WorkerOptions>, WorkerOptionsSetup>());
 
             services.TryAddEnumerable(ServiceDescriptor.Singleton<ILoggerProvider, WorkerLoggerProvider>());
-            services.AddSingleton(NullLogWriter.Instance);
-            services.AddSingleton<IUserLogWriter>(s => s.GetRequiredService<NullLogWriter>());
-            services.AddSingleton<ISystemLogWriter>(s => s.GetRequiredService<NullLogWriter>());
-            services.AddSingleton<IUserMetricWriter>(s => s.GetRequiredService<NullLogWriter>());
-            services.AddSingleton<FunctionActivitySourceFactory>();
+
+            services.TryAddSingleton(NullLogWriter.Instance);
+            services.TryAddSingleton<IUserLogWriter>(s => s.GetRequiredService<NullLogWriter>());
```

Comment:
For the core worker logger services, `TryAdd...` is sufficient, as this primarily wants to ensure those specific services are registered.